blog: SLSA end-to-end (#1458)
slsa-framework/slsa: docs/_posts/2025-07-16-slsa-e2e.md

---
title: Request for examples: SLSA End-to-End
author: Andrew McNamara, Tom Hennen
is_guest_post: false
---

## Goal

This document outlines a request for examples (RFE) for an end-to-end
implementation of the Supply-chain Levels for Software Artifacts
([SLSA](slsa.dev)) framework. The goal is to create a comprehensive
demonstration of how SLSA can be used to secure the software supply chain, from
source code to end-user consumption. These implementations will serve as a
reference for the community, showcasing best practices and providing a clear
adoption path for organizations looking to improve their software supply chain
security.

We’re looking for examples that document the entire end-to-end story and include
working implementations that people can try for themselves. We can then share

> these examples by hosting them on the SLSA blog or by linking to them elsewhere.

Reviewer: What would you anticipate organizations without an open source implementation would be able to demonstrate?

    of the code.

<p style="text-align: right">
Potential tools: <a href="https://github.com/slsa-framework/slsa-source-poc">slsa-source-poc</a>, <a href="https://gittuf.dev">gittuf</a></p>

### Build & Test

The build & test stage covers applying SLSA protections to the way software is
built and how the dependencies used in a build are managed. It leverages the
[SLSA Build track](/spec/v1.1/), and the draft
[Build Environment](/spec/draft/#build-environment-track) and
[Dependency](/spec/draft/#dependency-track) tracks.

Implementations should:

-   **Generate Build Provenance:** Generate SLSA build Level 3-compliant build
  provenance, which includes information about the build environment, the
  specific build steps, and the inputs and outputs of the build process.
-   **Generate Build Environment attestations**: (Optional) Generate SLSA BuildEnv
  L2-compliant BuildEnv attestations, which includes integrity information about
  the build environment at the start of the build.
-   **Generate an SBOM:** (Optional) Generate a Software Bill of Materials (SBOM)
  in a standard format (e.g., SPDX, CycloneDX) that lists all the components and
  dependencies of the built artifact.

> -   **Collect source VSAs for all used source dependencies:** For each dependency,
>   retrieve and store its source VSA to verify its origin and integrity.
>     -   **Optionally (additionally) verify VSA during build and fail build if
>     absent/doesn’t meet requirements:** Implement a build-time check to
>     [verify the validity of each dependency's source VSA](/spec/v1.2-rc1/verifying-source#how-to-verify-slsa-a-source-revision)
>     and fail the build if a VSA is missing or does not meet the defined security
>     policy.

Reviewer: Ah, of course. The use of the word `dependency` threw me off. We don't use the word dependency within the source track with this meaning. I think the sources would be represented as `resolvedDependencies` in the provenance. 

Does this make it any clearer or worse?

```suggestion
-   **Collect source VSAs for all used source resolvedDendencies:** For each source,
  retrieve and store its VSA to verify its origin and integrity.
    -   **Optionally (additionally) verify VSA during build and fail build if
    absent/doesn’t meet requirements:** Implement a build-time check to
    [verify the validity of each dependency's source VSA](/spec/v1.2-rc1/verifying-source#how-to-verify-slsa-a-source-revision)
    and fail the build if a VSA is missing or does not meet the defined security
    policy.
```

      by a trusted and authorized builder.
        -   **From the expected build entrypoint:** Verify that the build was
      initiated from the correct and expected entry point.
        -   **Each source dependency has:**
            -   **A valid source VSA:** Verify the integrity and provenance of each
        source dependency by
        [checking its VSA](/spec/v1.2-rc1/verifying-source#how-to-verify-slsa-a-source-revision).
            -   **(optionally) has a desired source level:** Verify that each source
        dependency meets the minimum SLSA source level defined in the security
        policy.
    -   **Check SBOM:** Check that an SBOM from the expected tooling exists.
    -   **Check vulnerability attestation and triage:** Verify that all identified
    vulnerabilities have been triaged and that the remediation status is
    acceptable according to the security policy.
-   **Issue VSA including:**
    -   **SLSA build level:** The VSA should include the SLSA build level that the
    artifact has achieved.
    -   **Minimum SLSA Source Level of all sources:** The VSA should include the

>     minimum SLSA source level of all the source code used in the build.

Reviewer: I feel like this might be better to leave out for now? We can create an issue for it to discuss it further as well.

If you have a dependency on a Python wheel, this statement is not saying anything about the source levels. If you consume an sdist instead, however, this requirement would come into effect.

Maybe that is intended? Should the source track provide a way to distinguish between these sources?

>     minimum SLSA source level of all the source code used in the build.

Reviewer: I don't think we came to a resolution. I don't think we have precisely defined language to differentiate between the sources code for an artifact's build vs its dependencies.

>     -   **Minimum SLSA Source Level of all sources:** The VSA should include the
>     minimum SLSA source level of all the source code used in the build.

Reviewer: How about something like this? It isn't resolved... if want to show how you capture it, then great.
```suggestion
    -   **Minimum SLSA Source Level of all sources:** The VSA should include the
    minimum SLSA source level of all the source code used in the build. This can
    optionally include any dependencies if they are built from source.
```

>     minimum SLSA source level of all the source code used in the build.

Author: Works for me! Thanks!

-   **Verify the VSA:** The tooling should automatically verify the VSA against
  the artifact to ensure its integrity and authenticity following
  [the SLSA guidance](/spec/v1.1/verification_summary#how-to-verify).
-   **Fail if VSA fails verification:** The tooling should fail and alert the user
  if the VSA does not pass the verification process, preventing the use of a
  potentially compromised artifact.

<p style="text-align: right">
Potential tools: <a href="https://github.com/carabiner-dev/drop">Drop</a>, <a href="https://github.com/slsa-framework/slsa-verifier">slsa-verifier</a></p>

## Extras

Some implementations may wish to go above and beyond what’s been requested.

That may include:

-   **Advanced policy enforcement**: providing a more nuanced ability to set
  policy for advanced users.
-   **Integration with other security tools**
-   **Support for additional attestation formats**
-   **Support for attestations from different producers:** The implementation may
  be able to consume and verify attestations from different producers, such as
  different builders, source control systems, and SBOM generators.

## Submission

> Submissions should document:

Reviewer: Added a suggestion below.

-   **How they work:** Explains the end-to-end workflow of the implementation.

> -   \*\*How they leverage existing standards: \*\*Note which standards or existing

Reviewer: You marked the conversation as resolved with a suggestion that hasn't been applied. :)

  projects are used in the submission.
-   **How can a user adopt this workflow themselves:** A step-by-step guide that
  allows users to set up and use the implementation in their own environment.
  This can link to existing documentation of any technology used.
-   **The policy management process for the software producer:** A document that
  explains how to create, manage, and enforce security policies.
-   **Gaps and future areas of improvement:** A document that identifies the
  current limitations of the implementation and suggests areas for future
  improvement.
-   **A working example:** A complete, working example of the implementation.

> To submit your example end-to-end implementations please either send a PR
> creating the blog post yourself
> ([example](https://github.com/slsa-framework/slsa/pull/1435)) or
> [create an issue](https://github.com/slsa-framework/slsa/issues) with the title
> “SLSA e2e: …”..

Reviewer: ```suggestion
To submit your example end-to-end implementations please either send a PR
creating the blog post yourself
([example](https://github.com/slsa-framework/slsa/pull/1435)) or
[create an issue](https://github.com/slsa-framework/slsa/issues) with the title
“SLSA e2e: …”. If your example is too large for either of these formats,
you may put the remaining walkthrough content in an external resource
(i.e. a git repository) and link to it.
```
